src/skalo/positioning.rs

```diff
@@ -0,0 +1,256 @@
+
+    // initialize variables to store the genome
+    let mut genome_seq: Vec<u8> = Vec::new();
+    let mut genome_name: String = "".to_string();
+
+    // set the reader for compressed or uncompressed files
+    let buf = get_reader(&file_path);
+    let mut reader = Reader::new(buf);
+
+    // initialize sequence counter
+    let mut sequence_count = 0;
+
+    // process records one by one
+    while let Some(record) = reader.next() {
+        // increment sequence counter
+        sequence_count += 1;
+        if sequence_count > 1 {
+            panic!("\nError: more than one sequence detected in the reference genome file.\n");
+        }
+
+        // unwrap record (contains name, sequence, and quality)
+        let record_ready = match record {
+            Ok(record) => record,
+            Err(_) => continue, // Skip on error
+        };
+
+        // process the genome sequence
+        genome_seq = record_ready
+            .seq()
+            .iter()
+            .copied()
+            .filter(|&byte| !byte.is_ascii_whitespace())
+            .map(|byte| byte.to_ascii_uppercase())
+            .collect();
+
+        // get the genome name
+        genome_name = record_ready.id().unwrap().to_string();
+
+        // only consider sequences long enough to have a k-mer
+        if genome_seq.len() >= k {
+            // extract k-mers (slices from Vec<u8>)
+            for n in 0..(genome_seq.len() - k + 1) {
+                // get slice of Vec<u8>
+                let kmer = &genome_seq[n..n + k];
+
+                // convert k-mer to u128
+                if let Some(kmer_encoded) = encode_vecu8_u128(kmer) {
+                    // Skip k-mers that have already overflowed
+                    if overflow_kmers.contains(&kmer_encoded) {
+                        continue;
+                    }
+
+                    // insert or update the k-mer in the HashMap
+                    let positions = kmer_map.entry(kmer_encoded).or_insert_with(Vec::new);
+                    if positions.len() < 3 {
+                        positions.push((n + k) as u32);
+                    }
+
+                    // if positions exceed 3, remove the k-mer and add it to the overflow set
+                    if positions.len() > 3 {
+                        kmer_map.remove(&kmer_encoded);
+                        overflow_kmers.insert(kmer_encoded);
+                    }
+                }
+            }
+        }
+    }
+
+    (kmer_map, genome_seq, genome_name)
+}
+
+// reader function
+fn get_reader(path: &PathBuf) -> Box<dyn BufRead + Send> {
+    let filename_str = path.to_str().unwrap();
+    let file = match File::open(path) {
+        Ok(file) => file,
+        Err(error) => panic!("Error opening file: {:?}.", error),
+    };
+
+    if filename_str.ends_with(".gz") {
+        Box::new(BufReader::new(MultiGzDecoder::new(file)))
+    } else {
+        Box::new(BufReader::new(file))
+    }
+}
```

Comment:
If refactoring the above this can probably be removed, but in any case it's easier just to use needletail